hazelcast/cp.py

```diff
@@ -152,3 +200,203 @@ def _get_group_id(self, proxy_name):
+
+            def clear(_):
+                self._sessions.clear()
+                self._mutexes.clear()
+                self._thread_ids.clear()
+
+            return combine_futures(futures).continue_with(clear)
+
+    def _request_generate_thread_id(self, group_id):
+        codec = cp_session_generate_thread_id_codec
+        request = codec.encode_request(group_id)
+        invocation = Invocation(request, response_handler=codec.decode_response)
+        self._context.invocation_service.invoke(invocation)
+        return invocation.future
+
+    def _request_close_session(self, group_id, session_id):
+        codec = cp_session_close_session_codec
+        request = codec.encode_request(group_id, session_id)
+        invocation = Invocation(request, response_handler=codec.decode_response)
+        self._context.invocation_service.invoke(invocation)
+        return invocation.future
+
+    def _get_or_crate_session(self, group_id):
+        with self._lock:
+            if self._shutdown:
+                error = HazelcastClientNotActiveError("Session manager is already shut down!")
+                return ImmediateExceptionFuture(error)
+
+            session = self._sessions.get(group_id, None)
+            if session is None or not session.is_valid():
+                with self._mutex(group_id):
+                    session = self._sessions.get(group_id)
+                    if session is None or not session.is_valid():
+                        return self._create_new_session(group_id)
+            return ImmediateFuture(session)
+
+    def _create_new_session(self, group_id):
+        f = self._request_new_session(group_id)
+        return f.continue_with(self._do_create_new_session, group_id)
+
+    def _do_create_new_session(self, response, group_id):
+        # called from the reactor thread only
+        response = response.result()
+        session = _SessionState(response["session_id"], group_id, response["ttl_millis"] / 1000.0)
+        self._sessions[group_id] = session
+        self._start_heartbeat_timer(response["heartbeat_millis"] / 1000.0)
+        return session
+
+    def _request_new_session(self, group_id):
+        codec = cp_session_create_session_codec
+        request = codec.encode_request(group_id, self._context.name)
+        invocation = Invocation(request, response_handler=codec.decode_response)
+        self._context.invocation_service.invoke(invocation)
+        return invocation.future
+
+    def _mutex(self, group_id):
+        mutex = self._mutexes.get(group_id, None)
+        if mutex:
+            return mutex
+
+        mutex = RLock()
+        current = self._mutexes.setdefault(group_id, mutex)
+        return current
+
+    def _start_heartbeat_timer(self, period):
+        if self._heartbeat_timer is not None:
+            return
+
+        def heartbeat():
+            if self._shutdown:
+                return
+
+            for session in list(self._sessions.values()):
```

Comment:
If you do something like `for session in self._sessions.values()`, then you cannot modify the `_sessions`, it would throw an error. What we do here is basically copying the values before the iteration and iterating over the copies

Reply:
Thanks. It sounds like we're safe here.